activerecord/test/cases/migration/exclusion_constraint_test.rb

```
@@ -0,0 +1,103 @@
+
+          constraint = exclusion_constraints.first
+          assert_equal "invoices", constraint.table_name
+          assert_equal "excl_rails_74c9160f55", constraint.name
+
+          assert_equal "daterange(start_date, end_date) WITH &&", constraint.expression
+        end
+
+        def test_added_exclusion_constraint_ensures_valid_values
+          @connection.add_exclusion_constraint :invoices, "daterange(start_date, end_date) WITH &&", using: :gist
+
+          Invoice.create(start_date: "2020-01-01", end_date: "2021-01-01")
+
+          assert_raises(ActiveRecord::StatementInvalid) do
+            Invoice.create(start_date: "2020-12-31", end_date: "2021-01-01")
+          end
+        end
+
+        def test_remove_exclusion_constraint
+          assert_equal 0, @connection.exclusion_constraints("invoices").size
+
+          @connection.add_exclusion_constraint :invoices, "daterange(start_date, end_date) WITH &&", using: :gist, name: "invoices_date_overlap"
+          assert_equal 1, @connection.exclusion_constraints("invoices").size
+          @connection.remove_exclusion_constraint :invoices, name: "invoices_date_overlap"
+          assert_equal 0, @connection.exclusion_constraints("invoices").size
+        end
+
+        def test_remove_non_existing_exclusion_constraint
+          assert_raises(ArgumentError) do
+            @connection.remove_exclusion_constraint :invoices, name: "nonexistent"
+          end
+        end
+      end
+    end
+  end
+else
+  module ActiveRecord
+    class Migration
+      class NoExclusionConstraintSupportTest < ActiveRecord::TestCase
+        setup do
+          @connection = ActiveRecord::Base.connection
+        end
+
+        def test_add_exclusion_constraint_should_be_noop
+          @connection.add_exclusion_constraint :test_exclusion_constraints, "daterange(start_date, end_date) WITH &&", using: :gist, where: "start_date IS NOT NULL and end_date IS NOT NULL", name: "date_overlap"
```

as above. i think these should fail loudly.